library/functions.lisp

```diff
@@ -111,12 +126,21 @@
     "Compute the complement of a unary Boolean function."
     (not (f x)))
 
+  (declare curry ((Tuple :left :right -> :result) -> :left -> :right -> :result))
+  (define (curry func left right)
+      (func (Tuple left right)))
+
   (declare uncurry ((:left -> :right -> :result) -> Tuple :left :right -> :result))
   (define (uncurry func tpl)
     (match tpl
       ((Tuple left right)
        (func left right))))
 
+  (declare pair-with ((:left -> :right) -> :left -> Tuple :left :right))
+  (define (pair-with func left)
```

Comment:
How about

    (pair-with left f right)

Also not sure this is the most useful, but I can see it.

Reply:
This is an extension of `iter:pair-with!` and can be used in general to make pairings `(x, f(x))`. This would allow any collection to do `(map (pair-with f) collection)`, including `iterator`, so `iter:pair-with!` could then be deprecated.